modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java

```diff
@@ -0,0 +1,412 @@
+        private final String gridName;
+
+        /** Regions. */
+        private final List<DataRegionConfiguration> regions = new ArrayList<>();
+
+        /** Caches. */
+        private final List<CacheConfiguration> caches = new ArrayList<>();
+
+        /** Default region configuration. */
+        @Nullable private DataRegionConfiguration dfltRegionConfiguration;
+
+        /**
+         * @param gridId Grid id.
+         */
+        ConfigurationBuilder(int gridId) {
+            this.gridName = getTestIgniteInstanceName(gridId);
+        }
+
+        /**
+         * @param regionName Region name.
+         * @param persistence Persistence.
+         */
+        ConfigurationBuilder withDefaultRegion(String regionName, boolean persistence) {
+            dfltRegionConfiguration = new DataRegionConfiguration()
+                .setName(regionName)
+                .setInitialSize(100L * 1024 * 1024)
+                .setMaxSize(500L * 1024 * 1024)
+                .setPersistenceEnabled(persistence);
+
+            return this;
+        }
+
+        /**
+         * @param regionName Region name.
+         * @param persistence Persistence.
+         */
+        ConfigurationBuilder withRegion(String regionName, boolean persistence) {
+            regions.add(new DataRegionConfiguration()
+                .setName(regionName)
+                .setInitialSize(100L * 1024 * 1024)
+                .setMaxSize(500L * 1024 * 1024)
+                .setPersistenceEnabled(persistence)
+            );
+
+            return this;
+        }
+
+        /**
+         * @param cacheName Cache name.
+         */
+        ConfigurationBuilder andCache(String cacheName) {
+            return andCache(cacheName, null);
+        }
+
+        /**
+         * @param cacheName Cache name.
+         * @param regionName Region name.
+         */
+        ConfigurationBuilder andCache(String cacheName, String regionName) {
+            caches.add(new CacheConfiguration().setDataRegionName(regionName).setName(cacheName));
+
+            return this;
+        }
+
+        /**
+         * This cache related with node via node filter.
+         *
+         * @param cacheName Cache name.
+         * @param regionName Region name.
+         */
+        ConfigurationBuilder andExclusiveCache(String cacheName, String regionName) {
+            caches.add(new CacheConfiguration()
+                .setNodeFilter(new NodeFilter(gridName))
+                .setDataRegionName(regionName)
+                .setName(cacheName)
+            );
+
+            return this;
+        }
+
+        /** Start node from builder */
+        public IgniteEx start() throws Exception {
+            IgniteConfiguration cfg = getConfiguration(gridName);
+
+            cfg.setConsistentId(gridName);
+
+            DataStorageConfiguration storageCfg = new DataStorageConfiguration();
+            storageCfg.setDataRegionConfigurations(regions.toArray(new DataRegionConfiguration[] {}));
```

Comment:
looks suspicious. maybe this way? v
```suggestion
            storageCfg.setDataRegionConfigurations(regions.toArray(newDataRegionConfiguration[regions.size()]));

```

Reply:
if (a.length < size) return (T[]) Arrays.copyOf(elementData, size, a.getClass());

It creates an array inside. Fixed it.

Reply:
For example, the IDEA suggests: .toArray(new DataRegionConfiguration[0])